Description: Use correct freeing function for JBIG2 images
 When jbig2_image_compose() errors out,
 remember to release all allocated pattern images.
 Previously the most recently allocated image would not be release.
 .
 Finally remember to free the array of images itself.
Origin: upstream, git://git.ghostscript.com/jbig2dec.git;a=commit;h=ade8a55
Author: Sebastian Rasmussen <sebras@gmail.com>
Bug: http://bugs.ghostscript.com/show_bug.cgi?id=703653
Forwarded: yes
Last-Update: 2021-04-22
---
This patch header follows DEP-3: https://dep.debian.net/deps/dep3/
--- a/jbig2_halftone.c
+++ b/jbig2_halftone.c
@@ -73,8 +73,10 @@
             new->patterns[i] = jbig2_image_new(ctx, HPW, HPH);
             if (new->patterns[i] == NULL) {
                 jbig2_error(ctx, JBIG2_SEVERITY_WARNING, JBIG2_UNKNOWN_SEGMENT_NUMBER, "failed to allocate pattern element image");
+                /* new->patterns[i] above did not succeed, so releasing patterns 0..i-1 is enough */
                 for (j = 0; j < i; j++)
-                    jbig2_free(ctx->allocator, new->patterns[j]);
+                    jbig2_image_release(ctx, new->patterns[j]);
+                jbig2_free(ctx->allocator, new->patterns);
                 jbig2_free(ctx->allocator, new);
                 return NULL;
             }
@@ -84,8 +86,10 @@
             code = jbig2_image_compose(ctx, new->patterns[i], image, -i * (int32_t) HPW, 0, JBIG2_COMPOSE_REPLACE);
             if (code < 0) {
                 jbig2_error(ctx, JBIG2_SEVERITY_WARNING, JBIG2_UNKNOWN_SEGMENT_NUMBER, "failed to compose image into collective bitmap dictionary");
-                for (j = 0; j < i; j++)
-                    jbig2_free(ctx->allocator, new->patterns[j]);
+                /* new->patterns[i] above succeeded, so release all patterns 0..i */
+                for (j = 0; j <= i; j++)
+                    jbig2_image_release(ctx, new->patterns[j]);
+                jbig2_free(ctx->allocator, new->patterns);
                 jbig2_free(ctx->allocator, new);
                 return NULL;
             }
